velero-plugin-for-aws/object_store.go

```diff
@@ -310,6 +358,11 @@ func (o *ObjectStore) ObjectExists(bucket, key string) (bool, error) {
 		Key:    aws.String(key),
 	}
 
+	if o.sseCustomerKey != "" {
+		req.SSECustomerAlgorithm = aws.String("AES256")
+		req.SSECustomerKey = &o.sseCustomerKey
+	}
```

Comment:
Not sure _why_ we need to configure this again since it's already been tackled in line 332. [1]

Reply:
Because in the case of SSE-S3 or KMS, the server takes care of your keys for you, so you don't have to supply a key to Get/Put/Head a file. However, with SSE-C the client supplies the key. You need this both for uploading files (`PutObject`), for downloading files (`GetObject`) and reading metadata (`HeadObject`). All encryption and decryption still happens on the server-side. The only difference is the way the key is supplied.

Reply:
Additional information can be found [here](https://docs.aws.amazon.com/AmazonS3/latest/userguide/ServerSideEncryptionCustomerKeys.html) and [here](https://docs.aws.amazon.com/AmazonS3/latest/userguide/specifying-s3-c-encryption.html). Specifically:

```
Amazon S3 rest APIs that support SSE-C
The following Amazon S3 APIs support server-side encryption with customer-provided encryption keys (SSE-C).

GET operation — When retrieving objects using the GET API (see [GET Object](https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectGET.html)), you can specify the request headers.
HEAD operation — To retrieve object metadata using the HEAD API (see [HEAD Object](https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectHEAD.html)), you can specify these request headers.
PUT operation — When uploading data using the [PUT Object](https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPUT.html) API (see PUT Object), you can specify these request headers.
```

Reply:
Do you know why with KMS way, we don't need to set the
```go
if o.kmsKeyID != "" {
    req.ServerSideEncryption = aws.String("aws:kms")
    req.SSEKMSKeyId = &o.kmsKeyID
}
```
in the `ObjectExists` and `GetObject` functions?

Reply:
Because the S3 server saves the encrypted data key for each file. It doesn't need the KMS id anymore at decryption time.

https://docs.aws.amazon.com/AmazonS3/latest/userguide/UsingKMSEncryption.html:
> If you choose to encrypt your data using a KMS key or customer managed key, AWS KMS and Amazon S3 perform the following actions:
> - Amazon S3 requests a plaintext [data key](https://docs.aws.amazon.com/kms/latest/developerguide/concepts.html#data-keys) and a copy of the key encrypted under the specified KMS key.
> -  AWS KMS generates a data key, encrypts it under the KMS key, and sends both the plaintext data key and the encrypted data key to Amazon S3.
> - Amazon S3 encrypts the data using the data key and removes the plaintext key from memory as soon as possible after use.
> - Amazon S3 stores the encrypted data key as metadata with the encrypted data.
> 
> When you request that your data be decrypted, Amazon S3 and AWS KMS perform the following actions:
> - Amazon S3 sends the encrypted data key to AWS KMS.
> - AWS KMS decrypts the key by using the same KMS key and returns the plaintext data key to Amazon S3.
> - Amazon S3 decrypts the ciphertext and removes the plaintext data key from memory as soon as possible.